syntax = "proto3";
package roundtrip;

import "google/protobuf/timestamp.proto";
import "google/protobuf/duration.proto";
import "google/protobuf/wrappers.proto";

// Message to test well-known protobuf types for round-trip conversion
message WellKnownTypesMessage {
  // Timestamp fields
  google.protobuf.Timestamp created_at = 1;
  google.protobuf.Timestamp updated_at = 2;
  google.protobuf.Timestamp expires_at = 3;
  
  // Duration fields
  google.protobuf.Duration timeout = 4;
  google.protobuf.Duration processing_time = 5;
  google.protobuf.Duration ttl = 6;
  
  // Optional well-known types
  optional google.protobuf.Timestamp optional_timestamp = 7;
  optional google.protobuf.Duration optional_duration = 8;
  
  // Repeated well-known types
  repeated google.protobuf.Timestamp event_timestamps = 9;
  repeated google.protobuf.Duration intervals = 10;
  
  // Map with well-known types as values
  map<string, google.protobuf.Timestamp> timestamp_map = 11;
  map<string, google.protobuf.Duration> duration_map = 12;
  
  // Wrapper types
  google.protobuf.DoubleValue double_value = 13;
  google.protobuf.FloatValue float_value = 14;
  google.protobuf.Int64Value int64_value = 15;
  google.protobuf.UInt64Value uint64_value = 16;
  google.protobuf.Int32Value int32_value = 17;
  google.protobuf.UInt32Value uint32_value = 18;
  google.protobuf.BoolValue bool_value = 19;
  google.protobuf.StringValue string_value = 20;
  google.protobuf.BytesValue bytes_value = 21;
}

// Edge cases for well-known types
message WellKnownEdgeCasesMessage {
  // Zero values
  google.protobuf.Timestamp zero_timestamp = 1;
  google.protobuf.Duration zero_duration = 2;
  
  // Max values
  google.protobuf.Timestamp max_timestamp = 3;
  google.protobuf.Duration max_duration = 4;
  
  // Negative duration (durations can be negative)
  google.protobuf.Duration negative_duration = 5;
  
  // Timestamp with nanosecond precision
  google.protobuf.Timestamp precise_timestamp = 6;
  
  // Duration with nanosecond precision
  google.protobuf.Duration precise_duration = 7;
}